### DISCUSSION TITLE
Upgrading events to zero or more than one event

### DISCUSSION THREAD

**Discussion by ppilev** [rasmus/EventFlow#859] [Q&A]
hello,
I have a question concerning upgrading to zero or more than one events. I've found the documentation says:

> **Be careful when working with event upgraders that return zero or more than one event, as this has an influence on the aggregate version and you need to make sure that the aggregate sequence number on upgraded events are valid in regard to the aggregate history.**

is there a recommended approach when dealing with such cases, i.e. historically I have the following scenario:
- emit event1 -> aggregate version 1 
- emit event2 -> aggregate version 2
- here I have registered an upgrader which upgrades event1 to produce zero events
- emit event3 -> `AggregateRoot.ApplyEvents` throws `InvalidOperationException` because the very first loaded event now is event2 and aggregate sequence number is 2.

thanks.

**Comment by rasmus** [comment]
You merely need to be very careful and test thoroughly. 

To edit the event on load, use the `IDomainEventFactory` to make a new version that has the correct meta data. The `EventUpgradeManagerTests` has a few tests that illustrate it.
